Relay Publish: Extend coverage

fbarbu15 opened the pull request:
Second iteration of the framework for interop/e2e tests:

- finished relay publish coverage: [26 tests](https://waku-org.github.io/waku-interop-tests/34/)
- extend waku node api with more actions
- add common custom logger lib
- add more reusable functions/steps for relay tests
- improve setup/cleanup hooks
- discovered multiple bugs in the process

fbarbu15 commented:
> I really like this! Just a comment that may or may not be helpful. Unless you specifically want to use `discv5`, I also think it is a good idea to connect nodes [manually](https://rfc.vac.dev/spec/16/#post_waku_v2_admin_v1_peers). It is an admin API method, so you have to enable it with `--rpc-admin=true`. I guess depending on the test case, this may be more handy than have to wait for the discovery protocol connections. Also, you get a bool with the success or failure of the connection, that you can use to keep going with the test or assert the error. Just a comment, of course.
> 
> Cool work!

Thanks for the feedback and the suggestion.
If it speeds up tests then I need to try it.
We could have dedicated tests for this connection method or have it behind a feature flag and enable/disable it for the whole suite. Need to think how it makes more sense, I tend to go with the 1st approach and have a dedicated suite for the Admin API. Either way it needs some representation for sure!

AlejandroCabeza commented on tests/relay/test_publish.py:
```diff
@@ -1,34 +1,231 @@
+        except Exception as ex:
+            assert "Peer node couldn't find any messages" in str(ex)
+
+    def test_publish_after_node_pauses(self):
```
Wouldn't this test imply a `check_published_message_reaches_peer` in between a `pause` and `unpause`?

fbarbu15 replied:
indeed the name is off, updating to test_publish_after_node_pauses_and_unpauses
having a `check_published_message_reaches_peer `in between a `pause `and `unpause `not sure if it makes sense because the connection is down when the node is in pause state

AlejandroCabeza replied:
Yes, I also felt that way at first when I implemented something similar in `nwaku`, but after actually running the test I discovered a behaviour:
* Relay can't receive stuff while paused
* Filter CAN receive stuff while paused

Although maybe it makes sense to just have that kind of test in the Filter suite.

fbarbu15 replied:
I actually tried it and it worked!
Node1 - started
Node2 - paused
Node1 - publishes message
Node2 - unpasues
Node2 - gets messages 
=> Message is received

I added a new test for this scenario. Thanks for the suggestion!

AlejandroCabeza replied:
No problem, it's my pleasure to give you more work 😂

AlejandroCabeza reviewed (commented):
LGTM! Just added a few small suggestions/improvements that you can pay attention to if you feel like so 😄

fbarbu15 commented on src/env_vars.py:
```diff
@@ -6,20 +6,22 @@
-    if env_var is not None:
-        print(f"{var_name}: {env_var}")
-    else:
+    if env_var in [None, ""]:
```
RUNNING_IN_CI is such a var

fbarbu15 commented on src/node/api_clients/base_client.py:
```diff
@@ -1,30 +1,30 @@
             response.raise_for_status()
         except requests.HTTPError as http_err:
-            logger.error("HTTP error occurred: %s", http_err)
+            logger.error("HTTP error occurred: %s. Response content: %s", http_err, response.content)
```
lazy logging since to be the better approach when comes to logging vs compared to f-strings
There are pros and cons and different opinions for both if you google it (ex https://medium.com/flowe-ita/logging-should-be-lazy-bc6ac9816906)
I've used both ways in the past, I don't have a personal preference. Most important thing is to have them consistent.

fbarbu15 replied:
I've updated all logs to use f-strings to be consistent with vars, Exceptions etc that use f-strings

AlejandroCabeza replied:
Oh, I didn't know! That's actually rather interesting! It hadn't occurred to me that the `*args` of the logging method would be chained that way! I saw the percentages and I assumed it was the typical (old) string format; good to know!

I take my comment back, both approaches have their advantages!

fbarbu15 commented on src/node/waku_node.py:
```diff
@@ -83,19 +87,52 @@ def stop(self):
-    @retry(stop=stop_after_delay(5), wait=wait_fixed(0.05), reraise=True)
+    def restart(self):
+        if self._container:
+            logger.debug("Restarting container with id %s", self._container.short_id)
```
Same as above

fbarbu15 commented on src/node/waku_node.py:
```diff
@@ -83,19 +87,52 @@ def stop(self):
         return self._api.info()
 
-    def set_subscriptions(self, pubsub_topics=[DEFAULT_PUBSUBTOPIC]):
+    def set_subscriptions(self, pubsub_topics=[DEFAULT_PUBSUB_TOPIC]):
```
Agree, updating as suggested, Thanks!

fbarbu15 commented on src/node/waku_node.py:
```diff
@@ -83,19 +87,52 @@ def stop(self):
+        elif self.is_gowaku():
+            return "gowaku"
+        else:
+            raise Exception("Unknown node type!!!")
```
Updating

fbarbu15 commented on src/steps/relay.py:
```diff
@@ -1,38 +1,83 @@
+    def check_published_message_reaches_peer(self, message, pubsub_topic=None, message_propagation_delay=0.1):
+        if not pubsub_topic:
+            pubsub_topic = self.test_pubsub_topic
+        self.node1.send_message(message, pubsub_topic)
```
Nice, didn't thinked at this, updating

fbarbu15 commented on tests/relay/test_publish.py:
```diff
@@ -1,34 +1,231 @@
+                self.node1.send_message(message, self.test_pubsub_topic)
+                success_content_topics.append(content_topic)
+            except Exception as ex:
+                assert "Bad Request" in str(ex) or "Internal Server Error" in str(ex)
```
Those are negative scenarios / error handling tests.
It's expected for them to fail. The test will fail when the exception will not be triggered

AlejandroCabeza replied:
Right, but why check the exception string contains either `Bad Request` or `Internal Server Error`? 

Shouldn't all (even ones that don't contain those strings) exceptions make the test fail?

fbarbu15 replied:
So in this case, when passing an invalid value in the request payload I would expect Bad Request and only pass the test in that case. However there are inconcistencies for now between how nwaku and gowaku behaven and we get "Internal Server Error" in some cases (I hope that this will be fixed soon). 
However there can be other error messages like Not Found, Request Timeout etc that I want to catch and fail the test in that case.
There is a case when a request crashed the container. I want to catch such cases. That's why I check the message string and only pass the tests if the error is what I;m expecting

fbarbu15 replied:
> Right, got it!
> 
> Note for the future: Might be handy to map those errors to their own exceptions.

Agreed, when both nodes behave the same, now it would complicate the code

fbarbu15 commented on src/libs/custom_logger.py:
```diff
@@ -0,0 +1,24 @@
+    return logLengthFilter()
+
+
+def get_custom_logger(name):
```
I just tried the `logging.config.dictConfig` now and to me my approach looks a little better as I don't have to definethe config as JSON (which increases the chanches to mess something UP)
Also there is no need for more config here because pytest contains the formatter and log level
So I would keep it as it is

AlejandroCabeza replied:
It's hard to mess it up because, as far as I recall, there's a parser/validator for the JSON. 

The only issue that was slightly bothering me was the "unnecessary" reconfiguration by repeatedly calling the method.

In any case, I'm not worried about this, I was just mainly wondering why the choice. If you feel comfortable this way then let's keep it :)

fbarbu15 commented:
> I really like this! Just a comment that may or may not be helpful. Unless you specifically want to use `discv5`, I also think it is a good idea to connect nodes [manually](https://rfc.vac.dev/spec/16/#post_waku_v2_admin_v1_peers). It is an admin API method, so you have to enable it with `--rpc-admin=true`. I guess depending on the test case, this may be more handy than have to wait for the discovery protocol connections. Also, you get a bool with the success or failure of the connection, that you can use to keep going with the test or assert the error. Just a comment, of course.
> 
> Cool work!

If it speeds up tests then I will 

> LGTM! Just added a few small suggestions/improvements that you can pay attention to if you feel like so 😄

Thanks for the suggestions Alex!
Implemented most of them

AlejandroCabeza commented on src/steps/relay.py:
```diff
@@ -1,38 +1,83 @@
+
+        assert received_message.payload == sent_message["payload"], assert_fail_message("payload")
+        assert received_message.contentTopic == sent_message["contentTopic"], assert_fail_message("contentTopic")
+        if "timestamp" in sent_message and sent_message["timestamp"] is not None:
```
Hahahaha, it does feel weird at first, I'll give you that 😛

AlbertoSoutullo commented:
Ok, I don't think I have anything else to add after that review.
You are in good hands with Python Alex's reviews.

romanzac reviewed (approved):
Florin, you are moving at great leaps forward. Thanks!

fbarbu15 commented:
> Florin, you are moving at great leaps forward. Thanks!

Thanks @romanzac ! Usually it starts fast but when it's mature enough it slows down
